src/urllib3/multipart/decoder.py

```diff
@@ -0,0 +1,149 @@
+"""Logic for parsing and decomposing a multipart response body."""
+from __future__ import annotations
+
+import email.parser
+import typing
+
+from .. import _collections
+from .. import response as _response
+from .encoder import encode_with
+
+
+class ImproperBodyPartContentError(Exception):
+    pass
+
+
+class NonMultipartContentTypeError(Exception):
+    pass
+
+
+def _header_parser(headers: bytes, encoding: str) -> typing.Sequence[tuple[str, str]]:
+    string = headers.decode(encoding)
+    items = email.parser.HeaderParser().parsestr(string).items()
+    items = typing.cast(typing.List[typing.Tuple[str, str]], items)
+    return items
+
+
+class BodyPart:
+    """This provides an easy way to interact with a single part of the body.
+
+    BodyParts include the headers in the part of the body as well as the body
+    content as bytes and optional converted to text based on the provided
+    encoding.
+
+    The encoding may be overridden by specifying ``part.encoding = '...'``.
+    """
+
+    def __init__(self, content: bytes, encoding: str):
+        #: Encoding used for the body part to decode body and headers
+        self.encoding = encoding
+        headers: dict[str, str] = {}
+        # Split into header section (if any) and the content
+        headerbytes, separator, bodybytes = content.partition(b"\r\n\r\n")
+        if b"\r\n\r\n" != separator:
+            raise ImproperBodyPartContentError("content does not contain CR-LF-CR-LF")
+
+        #: The bytes containing the body of this part
+        self.content = bodybytes
+        if headerbytes != b"":
+            headers.update(_header_parser(headerbytes.lstrip(), encoding))
+        #: The headers associated with this part
+        self.headers = _collections.HTTPHeaderDict(headers)
+
+    @property
+    def text(self) -> str:
+        """Content of the ``BodyPart`` in unicode."""
+        return self.content.decode(self.encoding)
+
+
+MD = typing.TypeVar("MD", bound="MultipartDecoder")
+
+
+class MultipartDecoder:
+    """This parses the full multipart/form-data payload.
+
+    The ``MultipartDecoder`` object parses the multipart payload of
+    a bytestring into a tuple of ``Response``-like ``BodyPart`` objects.
+
+    The basic usage is::
+
+        import requests
+        from requests_toolbelt import MultipartDecoder
+
+        response = request.get(url)
+        decoder = MultipartDecoder.from_response(response)
+        for part in decoder.parts:
+            print(part.headers['content-type'])
```

Comment:
Note to self: should these usage examples be updated to use urllib3 imports?

Reply:
This whole docstring might need rewriting – the phrase <code>``Response``-like ``BodyPart`` objects</code> only makes sense in a requests context, I think.

Reply:
This whole docstring might need rewriting – the phrase Response-like BodyPart objects only makes sense in a requests context, I think.

Reply:
Fixed!